feat: allow third party pre-game commands

AndASM opened the pull request:
The goal of this change is to allow other transformer executables to be called. Commands are required to be under `tools/cybercmd` and are executed with only a `-cybercmd` argument.

I am new to Rust and learning, so this code may be of... a quality level.

Add: third party command execution
Add: logging
Fix: toml parsing of partial config files

AndASM committed:
feat: allow third party pre-game commands

jac3km4 commented on src/lib.rs:
```diff
@@ -94,59 +113,129 @@ fn write_mod_cmd<W: Write>(config: &ModConfig, ctx: &ConfigContext, mut writer:
-                    .ok();
-                if *terminate_on_errors == Some(true) && !matches!(res, Some(st) if st.success()) {
-                    std::process::exit(0);
+        let mut tasks_done = TASKS_DONE.lock().unwrap();
```
I'm missing the point of the hash set and obtaining a lock for it inside the loop, are you trying to prevent this function from being executed from multiple threads and running concurrently? That shouldn't happen, but if you want to prevent that you could just lock on entering the function

AndASM replied:
The hash set is to prevent tasks from being run multiple times. This was inspired by my own addon which has a "InvokeScc" task to work around the bug where cybercmd requires a task to be defined to parse a toml file (it wont let one with just args.)

The mutex is an artifact of me not knowing much about hooking in third party processes, and learning rust. It appeared much of the code was being executed twice, possibly from different threads. So I used once_cell, and when I still ran into issues I wrapped it in a mutex. It may just be I am failing to understand something.

jac3km4 replied:
`try_get_final_cmd` should only be executed once, it's already behind a `sync::Lazy` so I don't think a mutex will make any difference
there was an issue with the hook running from multiple executables - that's why there's a `exe.eq_ignore_ascii_case("Cyberpunk2077")` check in the code, maybe that's what you were seeing, but I still don't think it should run the tasks twice

AndASM replied:
That makes a lot of sense. Could it be redscript itself? Since the InvokeScc code doesn't change the working directory, if the ASI loader is in any dll that gets loaded into the redscript process, it would also load the asi files?

Also the Cyberpunk loader / REDLoader could do it. But I'm running Cyberpunk2077.exe directly.

Whatever the case, that would explain some of the multiple execution behaviour I was seeing. How some of the statics were being initialized more than once, despite once_cell.

AndASM replied:
Ooo, or the error reporter executable. I can't recall if Windows loads proxies from the cwd. But it would make sense *not* to do that. That leaves one other executable in the cyberpunk folder, `REDEngineErrorReporter.exe`.

jac3km4 replied:
I think it was running from some process started by the game, IIRC you can find out by logging `std::env::current_exe()`

jac3km4 commented on src/lib.rs:
```diff
@@ -94,59 +113,129 @@ fn write_mod_cmd<W: Write>(config: &ModConfig, ctx: &ConfigContext, mut writer:
+                    if let Ok(cmd) = cmd {
+                        if cmd.starts_with(cybercmd_path) {
+                            let res = Command::new(&cmd)
+                                .arg("-cybercmd")
```
I feel like this flag maybe should be in the configuration, I don't think every executable necessarily needs to know what cybercmd is

AndASM replied:
As part of limiting execution of external commands to a specific folder, I figured the executables would probably be made specifically for it. It was an impulse based on the idea of trying to make it more secure. Your right though, I'm not sure it adds anything and handling configurable arguments may be better.

jac3km4 commented on src/lib.rs:
```diff
@@ -94,59 +113,129 @@ fn write_mod_cmd<W: Write>(config: &ModConfig, ctx: &ConfigContext, mut writer:
+                    dunce::simplified(&entry.path()).display(),
+                    match error.span() {
+                        Some(val) => format!("{:?}", val),
+                        None => "".to_string(),
```
I'd rather write String::new(), since it's statically guaranteed not to allocate

jac3km4 commented on src/lib.rs:
```diff
@@ -94,59 +113,129 @@ fn write_mod_cmd<W: Write>(config: &ModConfig, ctx: &ConfigContext, mut writer:
+                                .status()
+                                .ok();
+                            if !matches!(res, Some(st) if st.success()) {
+                                log.error(format!("Task \"{}\" failed when run.", task.command));
```
this logger is a bit awkward to use, normally a logger should accept `impl Display` rather than a `String` (it could then be wrapped with a macro that forwards `format_args!`)
but rather than writing a custom logger, I'd check for a small crate that can do file logging for us, I think there should be something

AndASM replied:
There are a few. I started with one of the premade loggers. I couldn't get it to work properly. So I wrote my own custom one to learn some more Rust.

Now in hindsight I think the logger was getting initialized twice, and that resulted in the file being overwritten.

I definitely have more enthusiasm than knowledge when it comes to Rust and updating cybercmd. 😂

jac3km4 replied:
I think there's an ordering issue between our entry point, and the initialization of the static rust logger, which screws up the default logger most libraries use
There might still be crates that expose a logger without having to use the static

jac3km4 replied:
I'll make a few comments on the logger, I think it can do with some minor adjustments

jac3km4 commented on src/paths.rs:
```diff
@@ -0,0 +1,15 @@
+    let game_path = match get_game_path() {
+        Ok(path) => path,
+        Err(e) => return Err(e),
+    };
```
why not
```suggestion
    let game_path = get_game_path()?;
```

jac3km4 commented on src/lib.rs:
```diff
@@ -94,59 +113,129 @@ fn write_mod_cmd<W: Write>(config: &ModConfig, ctx: &ConfigContext, mut writer:
 }
 
+pub fn log_dir() -> PathBuf {
+    let log_dir = (match paths::get_game_path() {
+        Ok(game_path) => game_path,
+        Err(e) => panic!("cybercmd cannot get game path: {}", e),
+    })
```
I think this is equivalent to
```suggestion
    let log_dir = paths::get_game_path().expect("cybercmd cannot get game path")
```
(expect includes formatted error by default too)

jac3km4 commented on src/lib.rs:
```diff
@@ -94,59 +113,129 @@ fn write_mod_cmd<W: Write>(config: &ModConfig, ctx: &ConfigContext, mut writer:
+    })
+        .join("r6")
+        .join("logs");
+    match fs::create_dir_all(log_dir.clone()) {
+        Ok(_) => log_dir,
+        Err(e) => panic!("cybercmd cannot get game path: {}", e),
+    }
```
same as above

jac3km4 replied:
++ you don't need to clone the `log_dir` if you pass it by reference to `create_dir_all`

AndASM replied:
I think it didn't work when I first tried. At that point I wasn't aware of the different Result and Error objects and probably messed up anyhow vs std or something similar.

I have some cleanup to do. Thanks for your comments! They've cleared up a lot for me. (Feel free to give more feedback if you'd like. It'd be helpful I'm sure. But I feel bad wasting your time on things I may already know how to fix from your other feedback!)

jac3km4 commented on src/logger.rs:
```diff
@@ -0,0 +1,100 @@
+
+use crate::paths;
+
+pub struct Logger {
+    file: Option<File>,
+    error: Option<Error>,
+    level: i8,
+}
```
I don't like the fact that this permits 4 states:
no logger and no error, both logger and error, only logger and no error, only error and no logger
more natural way to model it would be
```rs
pub struct Logger(InnerLogger);

enum InnerLogger {
  LogToFile(File),
  CannotInitialize(Error)
}
```
(`Logger` only being separated from `InnerLogger` to prohibit pattern matching)

jac3km4 commented on src/logger.rs:
```diff
@@ -0,0 +1,100 @@
+        }
+    }
+
+    pub fn log(&mut self, level: i8, message: String) -> Result<()> {
```
as mentioned above all log methods should probably accept `impl Display`, it's more convenient and doesn't force the caller to allocate strings - a value like `format_args!("write {0} to log", 1)` can be passed in without ever heap allocating

AndASM replied:
I'm going to see if I can get one of the premade logging crates to work. Probably one that uses the log facade, which brings with it macros like `log::info!(..)`. I think these should fulfil your suggestions, giving `format!`-like macros that take traits like Display and Debug.

AndASM committed:
wip: third-party commands with args

add: tasks take args array
add: tasks take parameters for args
remove: custom InvokeScc logic
add: InvokeScc exe path translation
note: redscript needs updated toml
pending: code cleanup and refactoring needed

AndASM committed:
fix: files missed in previous commit

AndASM committed:
change: Minor restructure xpath to root

AndASM committed:
feat: Use asi with test.exe

AndASM committed:
refactor: improve errors and split out code

AndASM committed:
fix: make Vortex installs work with toml files

Vortex will not install an asi with added files. This can be overridden with fomod xml.

AndASM committed:
chore: cleanup and dist redscript.toml

AndASM commented:
I may have split it out a bit much. The xtask may be a bit over-the-top, but it was good for learning.

I moved the "main" binary into a separate crate called "test", and made the library crate a workspace. I did this so I could change the binary to load the asi instead of statically linking the library. I found this useful for testing and troubleshooting.

I also created a "common" crate for code I used in the main library as well as xtask.

jac3km4 commented on crates/common/src/path.rs:
```diff
@@ -0,0 +1,47 @@
+
+#[macro_export]
+macro_rules! make_path {
+    ($first:expr, $($segments:expr),+) => {{
+        let mut path = common::PathBuf::new($first).expect("Invalid base path!");
+        $(path.push($segments);)*
+
+        path = {
+            if let Ok(normalized) = path.normalize() {
+                normalized
+            } else {
+                path.normalize_virtually().expect("Invalid path!")
+            }
+        };
+
+        // Automatically create directories to avoid errors and improve discoverability
+        // Don't try to create a directory named scc.exe
+        if path.extension().is_none() {
+            let _ = std::fs::create_dir_all(&path);
+        } else {
+            if let Ok(base_parent) = path.parent() {
+                if let Some(parent) = base_parent {
+                    let _ = std::fs::create_dir_all(&parent);
+                }
+            }
+        }
+
+        path
+    }}
+}
```
any reason why this isn't a function?
should probably be something like
```rs
fn make_path(arg: impl IntoIterator<Item=impl AsRef<Path>>) -> ...
```

AndASM replied:
Well, my understanding is that:
```rust
fn make_path(arg: impl IntoIterator<Item=impl AsRef<Path>>) -> ...
```
is syntactic sugar for:
```rust
pub fn make_path<I, P>(segments: I) -> ...
where
    I: IntoIterator<Item = P>,
    P: AsRef<Path>,
...
```

Which would make calls like: `make_path([&root, "target", "staging"])` invalid.

While each specific item in that call satisfies the bound `AsRef<Path>`, all instances of a generic parameter have to be the same type. So the bound type `P` becomes `&PathBuf` and then `"target"` causes a type mismatch error.

I believe this is part of the reason `std::fmt` presents functionality like `println` as macros, not as functions.

jac3km4 replied:
If you want to use a macro for that you should extract just that bit, e.g.
```
macro_rules! make_path {
    ($first:expr, $($segments:expr),+) => {{
        let mut path = common::PathBuf::new($first).expect("Invalid base path!");
        $(path.push($segments);)*
        path
    }}
}
```
instead of having everything in the macro

jac3km4 replied:
another minor note, but for good macro hygiene it's usually considered good practice to use explicit global scope for paths: `::common::PathBuf::new` instead of `common::PathBuf::new`

AndASM replied:
> use explicit global scope for paths: `::common::PathBuf::new` instead of `common::PathBuf::new`

Ah, I didn't know about the `::` prefix syntax. I thought I was already using global scope what with relative specifiers like `self::`, `super::` and `crate::` existing. Good to know, thanks!

jac3km4 replied:
the `::` bit is nice for macros, because the user could define their own `common` module and it'd lead to an ambiguity since macros are expanded wherever they're used; you don't need to worry about it outside of macros

AndASM replied:
> If you want to use a macro for that you should extract just that bit, e.g. `...` instead of having everything in the macro

So, you are saying that:
```rust
    let staging = make_path(make_path_join_macro!(&root, "target", "staging"));
```
instead of:
```rust
    let staging = make_path!(&root, "target", "staging");
```
would be the preferred way to call it?

jac3km4 replied:
you can extract the bit that does all the other stuff to a function and call it from the macro, most of the code inside the macro is monomorphic and could be put in a function (which will also make the life a bit easier for the compiler since it'll have less parsing and compiling to do)

jac3km4 replied:
the problem with big macros is they can easily explode code size, since it's basically just copy pasting tokens

jac3km4 commented on crates/test/Cargo.toml:
```diff
@@ -0,0 +1,10 @@
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
we don't need to keep the autogenerated comments

AndASM replied:
Following from this and learning from [rust_analyzer's](https://github.com/rust-lang/rust-analyzer) structure. I'm going to change the internal crates' manifests to version `0.0.0` and to use the workspace definitions e.g.
```toml
[package]
name = "..."
version = "0.0.0"

authors.workspace = true
edition.workspace = true
license.workspace = true
rust-version.workspace = true

[dependencies]
"..." = { workspace = true, ... }
```

jac3km4 commented on resources/installer/ModuleConfig.xml:
what is this?

AndASM replied:
It's an install file for the Vortex mod manager from nexusmods.com. You can customize the install, add user choices for installing different variants, etc.

I used it because Vortex's default Cyberpunk extension refuses to install an asi mod like cybercmd if it contains other mod files like the `r6/config/cybercmd` toml files. This let me bundle the updated redscript toml file with the zip in my tests. So I included the xml files as an option. (See the `cargo xtask dist` command and the related `xtask/src/dist.rs`)

AndASM replied:
As a side-note, I think you can also distribute C# based installers that are run in a partly-sandboxed C# scripting environment. Vortex has some interesting features. But I found the code and documentation a bit messy to dig through. (Rust is so nice for documentation and readable code!!)

jac3km4 replied:
it shows up as a binary file in Github for some reason

AndASM replied:
Oh interesting, it does. I'll have to take a look. Off the top of my head, I'd guess it probably has some non-UTF-8 text encoding that git isn't reading as text. I originally generated the files from a tool Nexus Mods recommends. I'll take a look and see about fixing that.

AndASM replied:
Yup, they were UTF-16le thanks to Windows' whole wide-character thing. Converting them to UTF-8 makes git (and thus GitHub) detect them properly.

jac3km4 commented on src/config.rs:
```diff
@@ -0,0 +1,104 @@
+#[derive(Clone, Default, Debug)]
+pub struct ConfigContext;
+
+impl microtemplate::Context for ConfigContext {
+    fn get_field(&self, field_name: &str) -> &str {
+        match field_name {
+            "game_dir" => PATHS.game_str.as_str(),
+            _ => "",
+        }
+    }
+}
```
this is a pretty odd/surprising solution, I think the struct used as the context should just contain a field for `game_dir`

AndASM replied:
I was trying to figure out the best place to process the `Task.values` HashMaps using microtemplates. Essentially I made it static. I've since learned a bit more rust syntax and a few more idioms that could solve the issue.

How would you go about it? Iterate through the values in an inner loop (or function) called from `get_configs` function and replace the values instead of doing it during deserialization? Convert ConfigContext to a static?

jac3km4 replied:
you can just initialize the `ConfigContext` whenever it's needed and pass it as a parameter,
I'd avoid statics unless they are necessary, like the `CMD_STR`'s purpose is to avoid re-running the initialization code, I don't think we need any other statics

jac3km4 commented on src/config.rs:
```diff
@@ -0,0 +1,104 @@
+    pub(crate) values: HashMap<String, String>,
+}
+
+fn deserialize_render_map<'de, D>(deserializer: D) -> Result<HashMap<String, String>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let hash_map = HashMap::<String, String>::deserialize(deserializer)?;
+    Ok(hash_map
+        .into_iter()
+        .map(|(key, value)| (key, render(value.as_str(), ConfigContext {})))
+        .collect())
+}
```
decoding shouldn't be coupled with templating, it'd be cleaner to just to standard deserialization and render the template later

AndASM replied:
This is related to the odd thing I did to ConfigContext. See my [comment](https://github.com/jac3km4/cybercmd/pull/2#discussion_r1142600016) there 🙂

jac3km4 commented on src/config.rs:
```diff
@@ -0,0 +1,104 @@
+        .collect())
+}
+
+impl microtemplate::Context for Task {
+    fn get_field(&self, field_name: &str) -> &str {
+        let result = if field_name == "game_dir" {
+            PATHS.game_str.as_str()
+        } else if let Some(val) = self.values.get(field_name) {
+            val.as_str()
+        } else {
+            ""
+        };
+        log::debug!("Task {} key {} = {}", self.command, field_name, result);
+
+        result
+    }
+}
```
same as above, I think the code should generally not implement `microtemplate::Context`, it's not the appropriate place to put app logic in

AndASM replied:
Okay, I can just copy (or reference?) game_dir instead of looking it up like that. Slowly getting my mind around borrows and references. It's an interesting set of abstractions Rust uses. It blurs a lot of the lines I have in my mind between low level assembly/C/C++ pointers and higher-level dynamic garbage-collected language's references and objects.

I like it. But old habits from different models keep tripping me up!

jac3km4 replied:
The simplest solution is to just shove a `PathBuf` into the struct. The way `microtemplate` works might then force you to do some redundant `clone`s, which would be a bit inefficient, but I think that'd be fine. Alternatively you could try to borrow it as a `Path`, but you don't have to if it proves difficult.

AndASM replied:
Oh, that's the `Task` `microtemplate::Context`. I have no idea how to implement that one using the derive macro. I don't think it can be. The keys are dynamic, they are defined in each task's toml file. So they can't be known to name as struct fields at compile time.

I can pre-set the game_dir value and remove that lookup. But the HashMap's `.get` call is not something microtemplate's derive macro supports.

jac3km4 replied:
What's the purpose of the hash map? I don't think it's being used anywhere, is the idea that the user can provide custom template parameters? If that's what you want then microtemplate would indeed not work well with this approach. Is that actually needed though? I'd rather just get rid of the hash map and only prepopulate it with known parameters like game_dir.

jac3km4 replied:
If the dynamic template parameters are needed, an alternative route that'd not be hard to implement is to have a generic DynamicKeys struct that only contains a HashMap and implements Context by just doing a lookup. Then, I'd delete the other implementations and just use the DynamicKeys with game_dir being inside of it in the HashMap. I think that's cleaner than having special cases for keys inside the implementations of Context.

AndASM replied:
> I think that's cleaner than having special cases for keys inside the implementations of Context.

Yup, that's pretty much what I meant with:

> I can pre-set the game_dir value and remove that lookup.

I'll look at taking it a step further with the custom DynamicKeys type. I'll see if I find a clean way to implement it around the deserialization and other processing.

> is the idea that the user can provide custom template parameters?

yes

> Is that actually needed though?

Not sure. When it comes to UX, ergonomics, and discoverability, I have trouble answering definitively "what is needed."

Here is the example toml I provided for redscript/scc. (Modified for clarity in this discussion.)
```toml
[[tasks]]
command = "scc"
args = ["-compile", "{path}", "-customCacheDir", "{custom_cache_dir}"]
path = "{game_dir}/r6/scripts"
custom_cache_dir = "{game_dir}/r6/cache/modded"
```
If you look at `args` you can see how `path` and `custom_cache_dir` are custom keys being defined for this task.

Alternatively it could be done plain like:
```toml
[[tasks]]
command = "scc"
args = ["-compile", "{game_dir}/r6/scripts", "-customCacheDir", "{game_dir}/r6/cache/modded"]
```

But I'm not sure that simplification is worthwhile for the difference in clarity. Other alternatives could be something like:
```toml
[[tasks]]
command = "scc"
"-compile" ="{game_dir}/r6/scripts"
"-customCacheDir" = "{game_dir}/r6/cache/modded"
```
Or, to avoid the quoting of keys:
```toml
[[tasks]]
command = "scc"
argument_prefix = "-"
compile ="{game_dir}/r6/scripts"
customCacheDir = "{game_dir}/r6/cache/modded"
```

Thoughts or preferences?

jac3km4 replied:
I prefer the plain version but I think options 1 and 2 are OK. 3 and 4 seem a bit too 'magical' and not obvious for an average user just trying to change the file.

AndASM replied:
> 1 and 2 are OK.

Okay, I'll continue with 1 and see if I can clean up the implementation to an acceptable level. It's easy enough to remove if 2 ends up being the final decision.

AndASM replied:
Oh, additional thought.

Side notes/commentary:
* I'm going to write this out as if you don't know the fine details of toml syntax. You may know it far better than I do. I think it works better to try to explain more clearly than less.
* I'm focusing on the toml format, since it explains and drives a lot of my decisions. Knowing the intended goal may help you better understand and evaluate my bizarre and chaotic work 😄 

Now back to the thought, the topic of this comment -
We could modify the syntax to something like:
```toml
[tasks.redscript]
command = "scc"
args = ["-compile", "{path}", "-customCacheDir", "{custom_cache_dir}"]

[tasks.redscript.arg_values]
path = "{game_dir}/r6/scripts"
custom_cache_dir = "{game_dir}/r6/cache/modded"
```

Or doing similar with array syntax combined with subtable syntax:
```toml
[[tasks]]
command = "scc"
args = ["-compile", "{path}", "-customCacheDir", "{custom_cache_dir}"]

[tasks.arg_values]
path = "{game_dir}/r6/scripts"
custom_cache_dir = "{game_dir}/r6/cache/modded"
```
Since a subtable in toml applies to the most recently defined array element, those argument values go in the above [[task]] element.

We can also use alternative equivalent syntax like dotted keys:
```toml
[tasks.redscript]
command = "scc"
args = ["-compile", "{path}", "-customCacheDir", "{custom_cache_dir}"]
arg_values.path = "{game_dir}/r6/scripts"
arg_values.custom_cache_dir = "{game_dir}/r6/cache/modded"
```
or an inline table:
```toml
arg_values = { path = "{game_dir}/r6/scripts", custom_cache_dir = "{game_dir}/r6/cache/modded" }
```
Both of which would deserialize identically to the earlier examples.

jac3km4 replied:
I'd probably name it something more explicit like `template_args` or `substitutions` or something like that.
The syntax I'd prefer is:
```toml
[[tasks]]
command = "scc"
args = ["-compile", "{path}", "-customCacheDir", "{custom_cache_dir}"]
substitutions.path = "{game_dir}/r6/scripts"
substitutions.custom_cache_dir = "{game_dir}/r6/cache/modded"
```
Splitting substitutions to another subtable looks nice but it could be surprising to someone not deeply familiar with TOML (especially the fact that the order matters).

jac3km4 commented:
The general file structure is a bit chaotic. Since there's a `crates` directory, I think we should opt for a standard workspace layout, meaning having no crates outside of `crate/A`, `crate/B` and so on - I'd suggest moving the top-level `src` to a crate and `xtask` as well.

jac3km4 commented on xtask/src/config.rs:
```diff
@@ -0,0 +1,63 @@
+use normpath::BasePathBuf;
+use once_cell::sync::Lazy;
+
+pub static PATHS: Lazy<Paths> = Lazy::new(Paths::new);
```
does this actually need a static? as far as I can see it could just be loaded at runtime and passed as a parameter to functions with no clear disadvantage

AndASM replied:
I think I was mimicking the pattern from `CMD_STR` without thinking about the reason (being an externally called pre-defined hooked function.)

Good catch. I'll go look at that again, fix it, and look for other artifacts of that mental pattern!

jac3km4 commented on xtask/src/config.rs:
```diff
@@ -0,0 +1,63 @@
+
+pub static PATHS: Lazy<Paths> = Lazy::new(Paths::new);
+
+pub struct Paths {
```
it seems like more a config, maybe `Config` would be a more appropriate name for this

AndASM replied:
It used to be `CONFIG.paths`, then I made it just PATHS because it was just paths... then I added other config... *sigh*

I really need to stop prematurely optimizing things. Especially if it's something like project structure. This is the same issue as the weird layout I ended up with because I moved the main cybercmd crate back to the root instead of keeping it in `crates/`.

AndASM commented:
> The general file structure is a bit chaotic. Since there's a `crates` directory, I think we should opt for a standard workspace layout, meaning having no crates outside of `crate/A`, `crate/B` and so on - I'd suggest moving the top-level `src` to a crate and `xtask` as well.

I originally had it like that. I had previously had issues with virtual manifest based workspaces in some of my earlier attempts at Rust. I arranged it like this on an impulse to clean it up. I've since read [some guidance from matklad](https://matklad.github.io/2021/08/22/large-rust-workspaces.html) which seems sound, that warns against doing it the way I did and suggests what you are suggesting. It suggests it due to problems which I have encountered. 😂

Time for a quick refactor / reorganization back to the virtual manifest!

AndASM committed:
refactor: Standard workspace structure

AndASM committed:
fix: Convert xml from UTF-16le to UTF-8 for git

AndASM committed:
fix: Remove unused legacy fields from info.xml

AndASM committed:
fix: macro split into internal function

AndASM committed:
fix: feedback from upstream pull req #2

AndASM commented:
I think I got it all. Mind taking another look?

jac3km4 commented on crates/common/src/file.rs:
```diff
@@ -0,0 +1,28 @@
+use zip::CompressionMethod;
+use zip_extensions::ZipWriterExtensions;
+
+pub fn download_file<S: Into<String>, P: AsRef<Path>>(source: S, dest: P) -> anyhow::Result<()> {
+    let mut response = reqwest::get(source.into())?;
```
`source` probably makes more sense to be `AsRef<str>` since we don't need an owned String here (reqwest::get doesn't need an owned String)

jac3km4 commented on crates/common/src/file.rs:
```diff
@@ -0,0 +1,28 @@
+    Ok(())
+}
+
+#[allow(deprecated)]
```
what's the deprecation?

AndASM replied:
Originally I was using the `zip` crate directly, rather than zip_extensions. Under `zip` all the `..._from_path(...)` functions are deprecated because they strip '..' from the path instead of handling it. I have opinions, but they are not relevant here 😆

Now it is a leftover I forgot to remove. That will be corrected in the next push.

jac3km4 commented on crates/cybercmd/src/config/app_context.rs:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    fn get_game_path() -> Result<PathBuf, PathsError> {
+        #[rustfmt::skip]
+            let game_path = std::env::current_exe()?
+            .normalize()?
+            .ancestors().nth(3).ok_or(PathsError::NoParent)?
+            .normalize_virtually()?;
```
why this `rustfmt::skip`? I'd just let it format

AndASM replied:
It was annoying me by breaking `.ancestors().nth(3).ok_or(PathsError::NoParent)?` into multiple lines. It didn't look pretty. Maybe not the best reason, but it is the answer to "why?"

jac3km4 replied:
I'd rather just go with whatever `rustfmt` does, the `let` keyword is also weirdly aligned here. You can maybe get it to keep it in one line by extracting an intermediate `let`.

jac3km4 commented on crates/cybercmd/src/config/argument_context.rs:
```diff
@@ -0,0 +1,43 @@
+
+impl microtemplate::Context for ArgumentContext {
+    fn get_field(&self, field_name: &str) -> &str {
+        let result = if let Some(val) = self.0.get(field_name) {
+            val.as_str()
+        } else {
+            ""
+        };
+        result
```
defaulting to an empty string is a bit meh, but there's no way to elegantly fail here, so I guess it's OK for now

I'd slightly change it though:
```suggestion
        self.0.get(field_name).map(AsRef::as_ref)
```

jac3km4 commented on crates/cybercmd/src/config/logging.rs:
```diff
@@ -0,0 +1,51 @@
+pub(super) fn setup_logging(paths: &Paths) -> Result<()> {
+    let window_roller = FixedWindowRoller::builder()
+        .build(
+            Path::join(Path::new(&paths.logs), "cybercmd{}.log")
```
maybe just use `.` syntax
```suggestion
            Path::new(&paths.logs).join("cybercmd{}.log")
```

AndASM replied:
That reminds me, I 'borrowed' that code from the documentation and should attribute it. Thus the different coding style. I'll fix that... if we keep that code.

I intend to switch the logger to `flexi_logger`.

I found another rust-based project called [redscript](https://github.com/jac3km4/redscript) in the CP2077 modding ecosystem by someone called "jekky" or "jac3km4". That project uses `flexi_logger` for logging. It outputs to the same log directory as the one I'm using. Assuming the logger is compatible with the asi dll, I thought it would be a better choice for consistency in both developer and user experience.

Especially since we are using the `log` facade that makes drop-in replacements trivial.

jac3km4 replied:
I like `flexi_logger` and `fern`, but `log4rs` works too, I don't have a strong preference here

AndASM replied:
Since neither of us has a strong preference for any logger, I copied `fn setup_logger(...)` from scc. It seemed like the easiest way to get a consistent experience. It also worked as a near drop-in replacement for my `fn setup_logging(...)`

jac3km4 commented on crates/cybercmd/src/config/logging.rs:
```diff
@@ -0,0 +1,51 @@
+            "[{d(%Y-%m-%d %H:%M:%S)}] {l:<5.5}: {m:}\n",
+        )))
+        .build(
+            Path::join(Path::new(&paths.logs), "cybercmd.log"),
```
same as above

jac3km4 commented on crates/xtask/src/config.rs:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    fn project_root() -> Result<PathBuf, PathsError> {
+        #[rustfmt::skip]
+            let root = BasePathBuf::new(env!("CARGO_MANIFEST_DIR"))?
```
same, I don't think `#[rustfmt::skip]` is needed

jac3km4 commented:
Not mandatory, but since this PR is big and the project didn't have any linting tool set up, I'd check if we can catch some issues with clippy.
This is the clippy setup I use in `redscript` (goes to `.cargo/config.toml`):
```toml
[target.'cfg(all())']
rustflags = [
    "-Wclippy::all",
    "-Wclippy::match_same_arms",
    "-Wclippy::semicolon_if_nothing_returned",
    "-Wclippy::single_match_else",
    "-Wclippy::redundant_closure_for_method_calls",
    "-Wclippy::cloned_instead_of_copied",
    "-Wclippy::redundant_else",
    "-Wclippy::unnested_or_patterns",
    "-Wclippy::unreadable_literal",
    "-Wclippy::type_repetition_in_bounds",
    "-Wclippy::equatable_if_let",
    "-Wclippy::implicit_clone",
    "-Wclippy::default_trait_access",
    "-Wclippy::explicit_deref_methods",
    "-Wclippy::explicit_iter_loop",
    "-Wclippy::inefficient_to_string",
    "-Wclippy::match_bool",
    "-Dwarnings",
    "-Drust_2018_idioms",
]
```

AndASM committed:
fix: more feedback from upstream pull req #2

change: switch log4rs to flexi_log
add: clippy lint groups
fix: assorted lint warnings
fix: cargo project detection
change: use build-std for binary size
change: single codegen-unit for release optimization
fix: calculate args before hook
change: duplicate xtask code to new staging module
fix: remove unnecessary skip and deprecated attributes
new: add file_name to GameConfig for error messages
fix: cwd usage for xshell in xtask
fix: remove unnecessary clones in favor of slices

AndASM commented:
> Not mandatory, but since this PR is big and the project didn't have any linting tool set up, I'd check if we can catch some issues with clippy.

I've added these settings (and some additional lint groups) to the cargo config. I have my editor set up to run clippy every time I save a file. Fortunately there were very few warnings to deal with as a result. Well, that and your feedback of course.

I've made the most recent set of changes. Take a look when you've got time.

jac3km4 reviewed (approved):
LGTM! good job making it through the scrupulous review

jac3km4 commented:
I think a good next step would be to add a CI action that runs clippy, produces a nexus archive and creates a release, tbh I am not sure how to do it yet with the new project structure, but I'll have a look when I've got time

AndASM commented:
> I think a good next step would be to add a CI action that runs clippy, produces a nexus archive and creates a release, tbh I am not sure how to do it yet with the new project structure, but I'll have a look when I've got time

Want any help? It seems I'm going to have extra time again this week. I can answer questions, give advice, or even draft the action (I'm assuming GitHub actions) if you want.


As for the basics. Are you familiar with the [cargo xtask convention](https://github.com/matklad/cargo-xtask)?

Short-ish version if you're not: it's not a binary, plugin, or library. It's a convention for a command alias. In this case, I've aliased the `cargo xtask` subcommand to `cargo run --package xtask --`. Cargo will automatically build xtask and download any required crates/libraries.

The convention for the `cargo xtask dist` subcommand is to put the packaged release artifacts in the `target/dist` subdirectory. I modelled my `crates/xtask/src/dist.rs` file after your `resources/package.ps1` file. It will build the release profile of the cybercmd crate, then package `cybercmd.zip` and `cybercmd-standalone.zip` in  `target/dist` with all the extra files and downloads.

AndASM commented:
> good job making it through the scrupulous review

Thank you so much for that! If only that kind of constructive feedback was normal in society in general. That was exceedingly helpful for my efforts to better learn the language. I think I got more out of that then I'd get out of your average university class.

Now I can break things in so many new and bizarre ways! Mwahaha!

jac3km4 commented:
If you want to make a CI Github action and submit a PR, I'm open to that.
I'm thinking to start with a simple clippy on build + automated release action that uploads the nexus archive.
